# INRIA ASSIGNEMENT

Le but de ce devoir est dans un premier temps la détection et la suppression d'entrées doublons dans une base de données puis dans un second temps l'exploration et la visualiation des données fournis

In [34]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine('sqlite:///data.db', echo=False)
con = engine.connect()
df_patient = pd.read_sql('select * from patient', con=con)
df_pcr = pd.read_sql('select * from test', con=con)
con.close()

Deux bases de données sont fournis : ***df_patient*** et ***df_pcr***.


In [35]:
df_patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
patient_id       20000 non-null int64
given_name       19560 non-null object
surname          19575 non-null object
street_number    19618 non-null float64
address_1        19204 non-null object
suburb           19788 non-null object
postcode         19801 non-null object
state            18010 non-null object
date_of_birth    17989 non-null float64
age              16003 non-null float64
phone_number     19081 non-null object
address_2        7893 non-null object
dtypes: float64(3), int64(1), object(8)
memory usage: 1.8+ MB


In [36]:
df_patient.head(5)

patient_id given_name    surname  street_number            address_1  \
0      221958    matisse     clarke           13.0          rene street   
1      771155     joshua     elrick           23.0         andrea place   
2      231932      alice     conboy           35.0     mountain circuit   
3      465838     sienna   craswell           39.0  cumberlegeicrescent   
4      359178     joshua  bastiaans          144.0        lowrie street   

          suburb postcode state  date_of_birth   age phone_number address_2  
0     ellenbrook     2527    wa     19710708.0  32.0  08 86018809  westella  
1   east preston     2074   nsw     19120921.0  34.0  02 97793152   foxdown  
2       prospect     2305   nsw     19810905.0  22.0  02 20403934      None  
3          henty     3620    wa     19840809.0  30.0  02 62832318    jodane  
4  campbell town     4051   nsw     19340430.0  31.0  03 69359594      None

In [37]:
df_pcr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8800 entries, 0 to 8799
Data columns (total 2 columns):
patient_id    8800 non-null int64
pcr           8800 non-null object
dtypes: int64(1), object(1)
memory usage: 137.6+ KB


In [38]:
df_pcr.head(5)

patient_id       pcr
0      653091         N
1      347667  Negative
2      708481  Positive
3      148825  Negative
4      150081  Negative

La première base de données est un dataframe contenant des entrées de patients (id,nom,prénoms,adresse,code postal). Certaines colonnes contiennent des entrées nulles. 
La seconde contient des id de patients avec les résultats au test PCR. Elle ne contient pas d'entrées nulles

On va donc commencer par "nettoyer" les deux dataframes.


## 1. Nettoyage du dataframe : ***df_patient***

In [39]:
df_patient.dtypes


patient_id         int64
given_name        object
surname           object
street_number    float64
address_1         object
suburb            object
postcode          object
state             object
date_of_birth    float64
age              float64
phone_number      object
address_2         object
dtype: object

<a id='incoherence_donnees'></a>
On regarde tout d'abord la qualité/fiabilité des données fournis. Cela passe notamment par les types des différentes colonnes d'entrées. 
- ***Patient ID*** : Le numéro ID d'un patient, normalement doit être unique, est en général de type entier. Ce qui est le cas ici
- ***given_name*** : Le prénom officiel du patient est une string donc categorical donc a bien comme *dtype* un objet. De même pour surname; adress_1,adress_2,suburb,state.

- ***phone_number*** : Le numéro de téléphone, peut contenir une extension correspondant à un état ou un espace donc il est préférable de le laisser en string

- ***date_of_birth*** : Les dates de naissances des patients pourraient être sous type Timestamp ou au mieux sous forme de int64, de même pour l'âge et le numéro de rue. 

- ***postcode***  : Le code postal est un nombre allant de 3 à 4 chiffres devrait être sous type int64 et non une categorical value. De plus les states fournis ne correspondent pas aux codes postaux




In [40]:
df_patient.describe()

patient_id  street_number  date_of_birth           age
count   20000.000000   19618.000000   1.798900e+04  16003.000000
mean   553918.938750      98.812468   1.949638e+07     27.534337
std    260234.961111    2481.908686   2.926916e+05      7.807008
min    100005.000000       0.000000   1.290041e+07      0.000000
25%    330194.250000      10.000000   1.924113e+07     24.000000
50%    556327.500000      24.000000   1.950032e+07     29.000000
75%    779419.500000      60.000000   1.975052e+07     33.000000
max    999993.000000  342951.000000   1.999123e+07     92.000000

D'un point de vue statistique, on constate des valeurs NULL pour les numéros de *rues, dob et l'âge*. Pour la date de naissance on remarque une date de naissance minimale à 1290. Enfin si on compare la moyenne d'âge avec l'âge médian on trouve une différence de 2ans. 

In [41]:
df_patient["age"].median()


29.0

In [42]:
len(df_patient["patient_id"].unique()) #Find how many uniques values in patient_id column

19798

On a 19798 ID de patients distincts au lieu de 20000, or l'ID de patient pour un hôpital est supposée être unique. Cela peut être dû soit à des doublons, soit les données regroupées dans cette base proviennent de différents hôpitaux qui ont chacun leur système numérique unique.

In [44]:
patient_id_dupes = list(df_patient["patient_id"].duplicated()) #Find the duplicates values in the patient_id column
patient_id_dupes_idx = [i for i in range(len(patient_id_dupes)) if patient_id_dupes[i]==True] #Retrieve their index

In [45]:
df_patient.loc[df_patient["patient_id"] == df_patient.loc[patient_id_dupes_idx[0],"patient_id"]]

patient_id given_name      surname  street_number           address_1  \
1456      119816        jed      rollins           63.0      mauldon street   
2377      119816        sam  iasenzaniro           12.0  booroondara street   

                suburb postcode state  date_of_birth  age phone_number  \
1456  macquarie fields     4169   nsw     19010805.0  NaN  07 73786737   
2377          brighton     6153    sa     19621207.0  NaN  03 90915909   

     address_2  
1456      None  
2377      None

Ces deux patients ont le même ID mais sont deux personnes différentes. On peut en conclure que l'ID de patient n'est pas un critère pour identifier un doublons dans notre base de données.

In [46]:
df_patient.describe(include='object')

given_name surname          address_1     suburb postcode  state  \
count       19560   19575              19204      19788    19801  18010   
unique       1171    4273               4326       2925     2021    102   
top        emiily   white  paul coe crescent  toowoomba     4350    nsw   
freq          373     539                 57        167       96   6143   

       phone_number       address_2  
count         19081            7893  
unique        17916            3460  
top     02 97010866  brentwood vlge  
freq              4              85

Pour les données non numériques, on remarque pour les états qu'on a 102 valeurs, or seule 10 états existent en Australie.

On va donc procéder la manière suivante pour le nettoyage de données :
<br>
<br>
    A) Remplissage des valeurs NULL pour les colonnes ***given_name, surname, adress_1, address_2, suburb, phone_number***. 
    Les données de ces colonnes non fournis ne peuvent être corrigés autrement.
    <br>
    <br>
    B) Correction des dates de naissances et âges
    <br>
    <br>
    C) Correction des codes postaux et des états

## A) Remplissage

In [47]:
#We work on a instance copy of the dataframe, thus any modifications we can always keep the original one without
#reloading it again
df_dedup = df_patient.copy()

In [48]:
df_dedup["given_name"]=df_dedup['given_name'].fillna('None')
df_dedup["surname"]=df_dedup['surname'].fillna('None')
df_dedup["address_1"]=df_dedup['address_1'].fillna('None')
df_dedup["address_2"]=df_dedup['address_2'].fillna('None')
df_dedup["phone_number"]=df_dedup['phone_number'].fillna(0)
df_dedup["street_number"]=df_dedup['street_number'].fillna(0)

In [49]:
df_dedup.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
patient_id       20000 non-null int64
given_name       20000 non-null object
surname          20000 non-null object
street_number    20000 non-null float64
address_1        20000 non-null object
suburb           19788 non-null object
postcode         19801 non-null object
state            18010 non-null object
date_of_birth    17989 non-null float64
age              16003 non-null float64
phone_number     20000 non-null object
address_2        20000 non-null object
dtypes: float64(3), int64(1), object(8)
memory usage: 1.8+ MB


## B) Correction de colonnes
### Etats et codes postaux

Comme on l'a souligné auparavant des [incohérences](#incoherence_donnees) apparraissent dans la correspondance des codes postaux et des états. L'Australie est composé de 8 états ayant chacun un intervalle de données et une abbréviation spécifique. 
***


- __Exemple:__
<br>
<br>
***Premier cas :**
On prend une entrée dans la liste au hasard. Le postcode écrit est *2428* ce qui correspond au state **NSW** or le state écrit est **QLD**. 
***Second cas :**
On prend une entrée dans la liste au hasard. Le state écrit est *vc* ce qui correspond à aucun state  officiellement mais se rapproche du state  **VIC**. 

In [50]:
print(" First Example : ")
print(df_dedup.iloc[15,6:8])
print("  ")
print(" Second Example : ")
print("  ")
print(df_dedup.iloc[301,6:8])


 First Example : 
postcode    2428
state        qld
Name: 15, dtype: object
  
 Second Example : 
  
postcode    4574
state         vc
Name: 301, dtype: object


Pour mettre à jours nos deux colonnes on effectue le procédé suivant : 

1. On commence par vérifier les valeurs NaN dans notre colonne **Postcode** ainsi que les entrées ne correspondant pas à un postcode.On fait pareillement pour les states. On leur assigne une valeur nulle. 

2. On définis un dictionnaire des states avec l'intervalle dans lequel le postcode correspondant se trouve. Pour les postcodes non nulles on assigne les bon states à partir de notre dictionnaire.

3. Pour les postocdes nulles on vérifie si un state est fournis. Dans le cas où le state est fournis on assigne le postcode correspondant. Dans le cas où le state n'est pas fournis, on peut vérifier à partir du suburb : https://jumpshare.com/v/V8Cu6szjbgAzdcMXWfw9# (des BDD sont disponibles sur internet). 

4. Dans le cas où ni le state, ni le postcode ni le surburb n'est fournis, il est préférable de drop l'entrée car le but étant d'étudier la prévalence du COVID en fonction des zones géographiques cet entrée ne sera pas utilisable. *(on constatera par la suite que nous n'arriverons pas à cet étape là, tout les suburb, postcodes et states sont disponibles dans notre dataframe)*

In [51]:
#Create two new column to update our postcode and state and keep the original to compare
df_dedup["postcode_updated"]=df_dedup["postcode"]
df_dedup["state_updated"]=df_dedup["state"]
df_dedup["suburb_updated"]=df_dedup["suburb"]

In [52]:
#List of index of NULL postcode and state
index_null_postcode = df_dedup[df_dedup['postcode'].isnull()].index.tolist()
index_null_state=df_patient[df_patient['state'].isnull()].index.tolist()

In [53]:
#We create a new column to find index of postcode who are different than digits 
df_dedup["postcode_str"]=df_dedup["postcode"].astype(str)
odd_postcode=[i for i in range(len(df_dedup["postcode_str"])) if len(df_dedup.loc[i,"postcode_str"])>4]

In [54]:
print(df_dedup.iloc[odd_postcode[0],5:8])


suburb                      3023
postcode    port noarlunga south
state                        tas
Name: 3976, dtype: object


Le postcode fournis est en réalité une banlieue en Australie situé en Australie du Sud soit un code postale entre 5000 et 6000 or le code postal fournis ici est 3023 correspondant à l'état Victoria. 
On fait donc le choix de ne garder pour ces cas là que le state fournis
<a id='suburb'></a>

In [55]:
#Because all the entry in are swapped we swapped them:
for i in odd_postcode:
    df_dedup.loc[i,"suburb_updated"]=df_dedup.loc[i,"postcode"]
    df_dedup.loc[i,"postcode_updated"]=df_dedup.loc[i,"suburb"]

In [56]:
#Verify that all the suburbs are updated 
for i in odd_postcode:
    print(df_dedup.loc[i,"suburb_updated"])
    

port noarlunga south
naremburn
toowoobma
brighton
oraneg
blacktown
forest hill
port lincoln
ashfield
forest hill
toowoomba
regents park
kelmscott
donnybrook


In [57]:
#We check every postcode assigned is 3-4 digits
odd_postcode_verif=[i for i in odd_postcode if len(df_dedup.loc[i,"postcode_updated"])>4]
odd_postcode_verif

[9718]

In [58]:
df_dedup.loc[9718,"postcode_updated"]

'467l0'

In [59]:
df_dedup.loc[9718,"postcode_updated"]=0

In [60]:
#We assign the rest the zero value

for i in index_null_postcode:
    df_dedup.loc[i,"postcode_updated"]=0
for i in index_null_state:
    df_dedup.loc[i,"state_updated"]=0

In [61]:
df_dedup["postcode_updated"]=df_dedup["postcode_updated"].astype(int)
df_dedup.dtypes

patient_id            int64
given_name           object
surname              object
street_number       float64
address_1            object
suburb               object
postcode             object
state                object
date_of_birth       float64
age                 float64
phone_number         object
address_2            object
postcode_updated      int64
state_updated        object
suburb_updated       object
postcode_str         object
dtype: object

In [62]:
#We create two dictionnaries one for the mapping from postcode to states, the other one from states to postcode
postal_codes = dict([(i,"nt") for i in range(800,900)]+
                    [(i,"nsw") for i in range(1000,1999)]+
                    [(i,"nsw") for i in range(2000,2600)]+
                    [(i,"nsw") for i in range(2619,2900)]+
                    [(i,"act") for i in range(2600,2619)]+
                    [(i,"act") for i in range(2900,2921)]+
                    [(i,"nsw") for i in range(2921,3000)]+
                    [(i,"vic") for i in range(3000,4000)]+
                    [(i,"qld") for i in range(4000,5000)]+
                    [(i,"sa") for i in range(5000,5800)]+
                    [(i,"wa") for i in range(6000,6800)]+
                    [(i,"tas") for i in range(7000,7800)]
                   )
state_codes ={"nt":800,"nsw":1000,"act":2600,"vic":3000,"qld":4000,"sa":5000,"wa":6000,"tas":7000}
#For the state to postcode we assign the first value of the postcode range

In [63]:
#First case we update the state based on the non-null postcode
idx_not_null_postcode = [i for i in range(df_dedup.shape[0]) if df_dedup.loc[i,"postcode_updated"]!=0]

In [64]:
#Mapping the states to the dictionnary postal_codes
for i in idx_not_null_postcode:
    df_dedup.loc[i,"state_updated"]=postal_codes[df_dedup.loc[i,"postcode_updated"]]


In [65]:
#Verify
df_dedup.iloc[:5,5:]

suburb postcode state  date_of_birth   age phone_number address_2  \
0     ellenbrook     2527    wa     19710708.0  32.0  08 86018809  westella   
1   east preston     2074   nsw     19120921.0  34.0  02 97793152   foxdown   
2       prospect     2305   nsw     19810905.0  22.0  02 20403934      None   
3          henty     3620    wa     19840809.0  30.0  02 62832318    jodane   
4  campbell town     4051   nsw     19340430.0  31.0  03 69359594      None   

   postcode_updated state_updated suburb_updated postcode_str  
0              2527           nsw     ellenbrook         2527  
1              2074           nsw   east preston         2074  
2              2305           nsw       prospect         2305  
3              3620           vic          henty         3620  
4              4051           qld  campbell town         4051

In [66]:
#Second case we update the null postcode based on the non-null state
idx_state_not_null=[]
for i in range(df_dedup.shape[0]):
    if df_dedup.loc[i,"state_updated"]!=0 and i not in idx_not_null_postcode:
        idx_state_not_null.append(i)

In [67]:
for i in idx_state_not_null:

    df_dedup.loc[i,"postcode_updated"]=state_codes[df_dedup.loc[i,"state_updated"]]

In [68]:
#Verify
df_dedup.iloc[95:100,5:]

suburb postcode state  date_of_birth   age phone_number  \
95    coombabah     None   vic     19351020.0  26.0  03 90265874   
96   palmerston     4069    wa     19451003.0   NaN  04 65877742   
97  harris park     3658   vic     19330804.0  21.0  02 04698047   
98   burringbar     7330    wa            NaN  36.0  03 94670748   
99     labrador     2782  None     19710301.0  30.0  04 83500862   

               address_2  postcode_updated state_updated suburb_updated  \
95            hazel hill              3000           vic      coombabah   
96                  None              4069           qld     palmerston   
97  ainslie nursing home              3658           vic    harris park   
98              moorilla              7330           tas     burringbar   
99                tongul              2782           nsw       labrador   

   postcode_str  
95         None  
96         4069  
97         3658  
98         7330  
99         2782

In [69]:
#Final case : the state and postcode of the entry is a null value
state_postcode_null = []
for i in range(df_dedup.shape[0]):
    if df_dedup.loc[i,"state_updated"]==0 and df_dedup.loc[i,"postcode_updated"]==0:
        state_postcode_null.append(i)

In [70]:
print("We have {} rows with no postcode and state".format(len(state_postcode_null)))

We have 17 rows with no postcode and state


In [71]:
#Like before we look at the suburbs
sb_list = [df_dedup.loc[i,"suburb"] for i in state_postcode_null]
set(sb_list)

{'alexandra bridge',
 'casino',
 'dapto',
 'downer',
 'elwood',
 'greenacre',
 'kings cross',
 'lake illawarra',
 'lower templestowe',
 'morven',
 'ngunnawal',
 'northmead',
 'salisbury',
 'wagaman',
 'winchelsea south',
 'winston hills',
 'yarrawonga'}

In [72]:
#The length of the set corresponds to the length of the suburbs list, all suburbs are unique.
#Because the list of the length is short we make a dictionnary
suburb_to_postcode = {"downer":2602,"wagaman":800,"kings cross":1340,"lake illawarra":2528,
                          "elwood":3184,
                          "greenacre":2190,
                          "salisbury":5108,
                          "northmead":2152,
                          "yarrawonga":3730,
                          "lower templestowe":3107,
                          "winston hills":2153,
                          "morven":4468,
                          "winchelsea south":3241,
                          "dapto":2530,
                          "alexandra bridge":6288,
                          "casino":2470,
                          "ngunnawal":2913}

In [73]:
#First we update the postcode from the suburb then the state from the postcode
for i in state_postcode_null:
    df_dedup.loc[i,"postcode_updated"]=suburb_to_postcode[df_dedup.loc[i,"suburb_updated"]]
    df_dedup.loc[i,"state_updated"]=postal_codes[df_dedup.loc[i,"postcode_updated"]]
 

In [74]:
df_dedup.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 16 columns):
patient_id          20000 non-null int64
given_name          20000 non-null object
surname             20000 non-null object
street_number       20000 non-null float64
address_1           20000 non-null object
suburb              19788 non-null object
postcode            19801 non-null object
state               18010 non-null object
date_of_birth       17989 non-null float64
age                 16003 non-null float64
phone_number        20000 non-null object
address_2           20000 non-null object
postcode_updated    20000 non-null int64
state_updated       20000 non-null object
suburb_updated      19788 non-null object
postcode_str        20000 non-null object
dtypes: float64(3), int64(2), object(11)
memory usage: 2.4+ MB


In [75]:
#Finally we fill the null suburb with the string "None"
index_null_suburb = df_dedup[df_dedup['suburb'].isnull()].index.tolist()
for i in index_null_suburb:
    df_dedup.loc[i,"suburb_updated"]="None"

In [76]:
df_dedup.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 16 columns):
patient_id          20000 non-null int64
given_name          20000 non-null object
surname             20000 non-null object
street_number       20000 non-null float64
address_1           20000 non-null object
suburb              19788 non-null object
postcode            19801 non-null object
state               18010 non-null object
date_of_birth       17989 non-null float64
age                 16003 non-null float64
phone_number        20000 non-null object
address_2           20000 non-null object
postcode_updated    20000 non-null int64
state_updated       20000 non-null object
suburb_updated      20000 non-null object
postcode_str        20000 non-null object
dtypes: float64(3), int64(2), object(11)
memory usage: 2.4+ MB


On peut maintenant passer à la correction des ages et date de naissances

### Age et Date de naissance

In [77]:
df_dedup['dob_updated']= df_dedup["date_of_birth"]
df_dedup['age_updated']= df_dedup["age"]

In [78]:
index_null_dob = df_dedup[df_dedup['dob_updated'].isnull()].index.tolist()

In [79]:
index_null_age = df_dedup[df_dedup['age_updated'].isnull()].index.tolist()

In [80]:
df_dedup["age_updated"]=df_dedup["age_updated"].fillna(0) #valeurs null
df_dedup["dob_updated"]=df_dedup["dob_updated"].fillna(0) #valeurs 

In [81]:
df_dedup['dob_updated']=df_dedup['dob_updated'].apply(lambda n: n if n%1 else int(n)) #remove trailling 0
df_dedup['age_updated']=df_dedup['age_updated'].apply(lambda n: n if n%1 else int(n)) #remove trailling 0

In [82]:
df_dedup["dob_updated"] = [(str(i)) for i in df_dedup["dob_updated"]]
df_dedup["dob_updated"] = [int(i[:4]) for i in df_dedup["dob_updated"]]

In [83]:
odd_year=df_dedup[(df_dedup["dob_updated"]<1900) & (df_dedup["dob_updated"]>0) ].index.tolist()

In [84]:
df_dedup.iloc[8882] 

patient_id                          258272
given_name                            jack
surname                             jessup
street_number                          509
address_1                      stone place
suburb                              batlow
postcode                              None
state                                  qld
date_of_birth                  1.29004e+07
age                                     34
phone_number                   08 77804659
address_2           mayflower retrmnt vlge
postcode_updated                      4000
state_updated                          qld
suburb_updated                      batlow
postcode_str                          None
dob_updated                           1290
age_updated                             34
Name: 8882, dtype: object

In [85]:
df_dedup.loc[8882,"dob_updated"]=2020-df_dedup.loc[8882,"age_updated"]

In [86]:
dob_zero = [i for i in range(df_dedup.shape[0]) if df_dedup.loc[i,"dob_updated"]==0 ]
age_zero= [i for i in  range(df_dedup.shape[0] )if df_dedup.loc[i,"age_updated"] ==0 ]
age_dob_zero =[i for i in dob_zero if i  in age_zero]
for i in age_dob_zero:
    df_dedup.loc[i,"dob_updated"]=1950
    df_dedup.loc[i,"age_updated"]=70

In [87]:
dob_zero_not_age =[i for i in dob_zero if i not in age_zero]
for i in dob_zero_not_age:
    df_dedup.loc[i,"dob_updated"]=2020-df_dedup.loc[i,"age_updated"]

In [88]:
dob_zero_verif = [i for i in range(df_dedup.shape[0]) if df_dedup.loc[i,"dob_updated"]==0 ]
age_zero_verif= [i for i in  range(df_dedup.shape[0] )if df_dedup.loc[i,"age_updated"] ==0 ]

In [89]:
df_dedup["age_final"]=df_dedup["age_updated"]

In [90]:
for i in range(df_dedup.shape[0]):
    df_dedup.loc[i,"age_final"]=2020-df_dedup.loc[i,"dob_updated"]

In [91]:
#Verify
df_dedup.iloc[1]

patient_id                771155
given_name                joshua
surname                   elrick
street_number                 23
address_1           andrea place
suburb              east preston
postcode                    2074
state                        nsw
date_of_birth        1.91209e+07
age                           34
phone_number         02 97793152
address_2                foxdown
postcode_updated            2074
state_updated                nsw
suburb_updated      east preston
postcode_str                2074
dob_updated                 1912
age_updated                   34
age_final                    108
Name: 1, dtype: object

In [92]:
cols_to_drop = ["suburb","postcode","state","date_of_birth","age","postcode_str","age_updated"]

In [93]:
df_dedup_clean = df_dedup.copy()

In [94]:
df_dedup_clean = df_dedup_clean.drop(["suburb","postcode","state","date_of_birth","age","postcode_str","age_updated"], axis = 1) 

In [95]:
df_dedup_clean

patient_id given_name      surname  street_number  \
0          221958    matisse       clarke           13.0   
1          771155     joshua       elrick           23.0   
2          231932      alice       conboy           35.0   
3          465838     sienna     craswell           39.0   
4          359178     joshua    bastiaans          144.0   
5          744167         ky        laing          448.0   
6          210268    matthew        laing           11.0   
7          832180       jack      renfrey           27.0   
8          154886      adele         ryan           76.0   
9          237337   breeanne        wynne           12.0   
10         621430    michael        white           17.0   
11         187498       None        mason           20.0   
12         999159     joshua      bristow           21.0   
13         275203        max      godfrey          166.0   
14         291335    mikayla     paterson           20.0   
15         378167       cain      scarles           58.0   
16         141090    georgia        white          117.0   
17         429365       None       coffey         1409.0   
18         164820        lia        wilde           30.0   
19         887201   mitchekl  stephenskon           58.0   
20         421879     callum      nicolle            2.0   
21         818315       ruby      grafton           97.0   
22         427069   mystique        white           85.0   
23         876997        zac    rosendale            5.0   
24         299041   harrison      neumann           60.0   
25         891743    charlie     mcgregor           36.0   
26         203355      clain      kinnane            3.0   
27         335796       jake       clarke           66.0   
28         314176      talia       mitton           10.0   
29         284779  alexandra        miles           23.0   
...           ...        ...          ...            ...   
19970      489925     alicia     morrison            5.0   
19971      186491   takeisha     rawlings           37.0   
19972      144588      jamie     chandler          128.0   
19973      301657   danielle      newport          611.0   
19974      199152     connor        whait          563.0   
19975      752160   ashleigh         None           30.0   
19976      513727       kane     costigan            1.0   
19977      896551      maxin         None            4.0   
19978      438037   nicholas       jacura            6.0   
19979      116167     thomas     matthews            6.0   
19980      811863      ethan      statham           17.0   
19981      210782       hana      barazda           18.0   
19982      581005    rebecca        santi           13.0   
19983      230775   victoria     whiteley           12.0   
19984      134889     xavier       priest           28.0   
19985      407822      sarah     matthews           58.0   
19986      326406      kiara        cheel           89.0   
19987      684550     robert   makovetsky            1.0   
19988      235753       erin     jeffries            2.0   
19989      704579    michael     eglinton          215.0   
19990      690296     bayden          due         1084.0   
19991      871835    william     villiers           20.0   
19992      329489      amber  leeuwenburg           50.0   
19993      506906     cooper     sherriff           25.0   
19994      912344     fraser      mcneill           10.0   
19995      368617       abby         None           19.0   
19996      796658       dale       priest           19.0   
19997      511885    alessia       mawlai            4.0   
19998      985932     hannah        mason            1.0   
19999      925217       jack      mcveigh            5.0   

                      address_1 phone_number                address_2  \
0                   rene street  08 86018809                 westella   
1                  andrea place  02 97793152                  foxdown   
2              mountain circuit  02 20403934                     No

In [96]:
df_dedup_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
patient_id          20000 non-null int64
given_name          20000 non-null object
surname             20000 non-null object
street_number       20000 non-null float64
address_1           20000 non-null object
phone_number        20000 non-null object
address_2           20000 non-null object
postcode_updated    20000 non-null int64
state_updated       20000 non-null object
suburb_updated      20000 non-null object
dob_updated         20000 non-null int64
age_final           20000 non-null int64
dtypes: float64(1), int64(4), object(7)
memory usage: 1.8+ MB


On fait le choix de garder ces colonnes pour nous permettre d'avoir un choix multiples en terme de critères de sélection pour la détection des doublons dans un dataframe.

In [97]:
#Export dataframe to csv
#df_dedup_clean.to_csv('df_dedup_clean.csv',index=False)

## 1. Nettoyage du dataframe : ***df_pcr***

In [98]:
df_pcr.describe()

patient_id
count    8800.000000
mean   554175.918977
std    259491.959500
min    100037.000000
25%    331553.000000
50%    556969.000000
75%    778063.750000
max    999993.000000

In [99]:
print("On a donc {} ID de patients unique et {} doublons d'ID de patients".format(len(df_pcr["patient_id"].unique()),len(df_pcr)-len(df_pcr["patient_id"].unique())))


On a donc 8765 ID de patients unique et 35 doublons d'ID de patients


In [100]:
df_pcr[df_pcr.duplicated(keep=False)].sort_values(by="patient_id")

patient_id       pcr
169       159500         N
2934      159500         N
1473      227192         N
8249      227192         N
3251      328071  Positive
4741      328071  Positive
4939      411955         N
3569      411955         N
1265      465871         N
7254      465871         N
7056      476826         N
6395      476826         N
6176      483152  Negative
5505      483152  Negative
2164      512755  Negative
5745      512755  Negative
4078      527530  Negative
4859      527530  Negative
4666      651216         N
3767      651216         N
224       655839  Negative
5071      655839  Negative
5724      670706  Positive
1790      670706  Positive
2251      728889  Negative
410       728889  Negative
7586      848423         N
8492      848423         N

Les 35 doublons d'ID de patients ont les même résultats PCR, on peut donc les supprimer. 

In [101]:
df_pcr = df_pcr.drop_duplicates(keep=False)
df_pcr.sort_values(by="patient_id")

patient_id       pcr
1900      100037         N
3919      100215  Negative
1733      100294         N
1672      100363         P
1880      100404         N
6336      100727  Negative
3699      100882  Positive
4592      100917  Negative
4703      101051  Negative
6005      101082  Positive
5869      101095  Negative
4707      101577         N
8079      101832  Negative
453       101845         N
5834      101917         N
7450      101919         P
6393      102016  Negative
3825      102159         N
8095      102187  Positive
724       102526         N
1579      102631         N
2470      102866  Positive
5593      102921         N
5646      103174  Negative
5343      103196         N
5460      103212         P
8794      103358         N
5649      103599         N
8693      103631  Negative
7202      103681  Positive
...          ...       ...
7640      997691  Negative
1724      997849         N
4649      997928  Positive
6758      998035  Positive
2268      998120  Negative
6738      998138         P
5363      998196  Negative
4567      998217         N
281       998320  Negative
5934      998332         N
3327      998342  Negative
1106      998407  Negative
8703      998436  Positive
8308      998607  Positive
440       998628         N
5478      998633  Positive
1626      998662  Negative
8412      998792  Negative
1361      998860  Negative
2255      998902  Positive
5669      999005  Positive
4259      999083  Negative
463       999084         N
3027      999159  Positive
7753      999434         N
8770      999482         N
3453      999497  Positive
8010      999573  Negative
8326      999885  Negative
1965      999993  Negative

[8772 rows x 2 columns]

In [102]:
df_pcr.describe(include=['object'])

pcr
count   8772
unique     4
top        N
freq    3468

Pour le résultat du PCR on a donc 4 valeurs possibles : Negative, Postive, N ou P. Ces 4 valeurs sont similaires on peut donc les uniformiser pour avoir uniquement Positive et Negative, la colonne PCR renvoyant une valeur soit True(Positive,1) soit False (Negative,0)

In [103]:
bool_pcr = {"Negative":False, "N":False, "Positive":True, "P":True}

In [104]:
df_pcr_clean = df_pcr.copy()

In [105]:
df_pcr_clean =df_pcr_clean.replace({"pcr":bool_pcr})

In [106]:
df_pcr_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8772 entries, 0 to 8799
Data columns (total 2 columns):
patient_id    8772 non-null int64
pcr           8772 non-null bool
dtypes: bool(1), int64(1)
memory usage: 145.6 KB


In [107]:
df_pcr_clean.head(5)

patient_id    pcr
0      653091  False
1      347667  False
2      708481   True
3      148825  False
4      150081  False

In [108]:
#Export dataframe to csv
#df_pcr_clean.to_csv('df_pcr_clean_final.csv',index=False)

In [109]:
#Merge two dataframe based on the patient ID key
df_patient_pcr = pd.merge(df_pcr_clean,df_dedup_clean,on="patient_id",how="inner")
#We want only the patient who were tested, so we only keep rows where the patient id value exists in both pcr and
#patient dataframe

In [110]:
df_patient_pcr

patient_id    pcr given_name      surname  street_number  \
0         653091  False     daniel     campbell           58.0   
1         347667  False  sebastian      mchenry           12.0   
2         708481   True      sarah       rellos           20.0   
3         148825  False      chloe       brammy          238.0   
4         150081  False    charlie         None            6.0   
5         876796  False  alexander        dixon           11.0   
6         169224  False       kyle        eyles           79.0   
7         755915   True      jamie      krollig           72.0   
8         374486  False    rachael      wilkins           17.0   
9         634003  False     lauren     braunack            2.0   
10        918386  False    georgia      allanby           31.0   
11        466326   True     sophie     schuster           36.0   
12        605659  False     sophie      coleman           15.0   
13        400572  False       jack    crosswell            3.0   
14        631740  False     alexis      whillas           14.0   
15        884141   True     taylah      collier          320.0   
16        440459   True     bailey       hislop           19.0   
17        130447  False  madeleine      seppelt          200.0   
18        916386  False      corie         webb           91.0   
19        317285  False     phoebe        sarah           21.0   
20        982876  False      julia    humphreys           29.0   
21        780215  False      dylan     shepherd           72.0   
22        578366  False      james       tiller           18.0   
23        651250   True      aidyn     fernando           26.0   
24        781669   True  gabrielle          ang           16.0   
25        455612  False     jordan      bartone            7.0   
26        948689   True     claire         hand           46.0   
27        988704  False    brayden         hope            1.0   
28        455249  False       hudd       thomas           16.0   
29        294243  False     marcus        green           41.0   
...          ...    ...        ...          ...            ...   
8882      667256   True       kane     shepherd          309.0   
8883      589415  False       zali     cirvydas           21.0   
8884      783889  False     annika       nguyen            9.0   
8885      662080  False   ellouise         hope           70.0   
8886      942549  False     philip     alderman           81.0   
8887      313456  False  stephanie    muthukuda           42.0   
8888      116375   True   isabella         None            6.0   
8889      116375   True     emiily   piorkowski           12.0   
8890      450500  False   mcgregor        abbey          105.0   
8891      723944  False     hannah   pettingill           57.0   
8892      682393  False        sam      coltman           51.0   
8893      872592   True    juliana    poppleton            1.0   
8894      618534  False    brinley      coulson            5.0   
8895      580674  False   tanyshah         webb           34.0   
8896      165488  False       None         hope          105.0   
8897      714134  False      jaime     mcgregor           69.0   
8898      541157   True    jacinta       feeley           29.0   
8899      400589  False       ella       weetra           26.0   
8900      375478  False       abby  smallacombe           62.0   
8901      326884  False   angelica        dolan            6.0   
8902      213604  False       lucy     matthews           14.0   
8903      800336  False  alexandra       clarke           28.0   
8904      270457  False       lily      jimenea           34.0   
8905      658692  False      holly     dumbrell            1.0   
8906      103358  False   nicholas     hingston          100.0   
8907      249457  False    tiffany       loulps           31.0   
8908      279689   True     daniel       excell           58.0   
8909      975612  False    jessica       nguyen           10.0   
8910      884429   True    chelsea      coltma

In [111]:
#Export dataframe to csv
#df_pcr_clean.to_csv('df_patient_pcr_clean.csv',index=False)